packages/next/src/server/lib/router-utils/mcp.ts

```diff
@@ -0,0 +1,585 @@
+    case 'conflict':
+      return []
+    default:
+      invariant(route, (r) => `Unknown route type: ${r.type}`)
+  }
+}
+
+function arrayOrSingle<T>(value: T | T[]): T[] {
+  return Array.isArray(value) ? value : [value]
+}
+
+interface ModuleReference {
+  module: Module
+  export: string
+  chunkingType: string
+}
+interface Module {
+  /// The identifier of the module, which is a unique string.
+  ident: string
+  /// The path of the module. It's not unique as multiple modules can have the same path.
+  path: string
+  /// The distance to the entries of the module graph. Use this to traverse the graph in the right direction.
+  depth: number
+  /// The size of the source code of the module in bytes.
+  size: number
+  /// The size of the whole subgraph behind this module in bytes.
+  retainedSize: number
+  /// The modules that are referenced by this module.
+  references: ModuleReference[]
+  /// The modules that reference this module.
+  incomingReferences: ModuleReference[]
+}
+
+function createModuleObject(rawModule: NapiModuleInfo): Module {
+  return {
+    ident: rawModule.ident,
+    path: rawModule.path,
+    depth: rawModule.depth,
+    size: rawModule.size,
+    retainedSize: rawModule.retainedSize,
+    references: [],
+    incomingReferences: [],
+  }
+}
+
+function processModuleGraphSnapshot(
+  moduleGraph: NapiModuleGraphSnapshot,
+  modules: Module[],
+  entries: Module[]
+) {
+  const queryModules = moduleGraph.modules.map(createModuleObject)
+  for (let i = 0; i < queryModules.length; i++) {
+    const rawModule = moduleGraph.modules[i]
+    const queryModule = queryModules[i]
+
+    queryModule.references = rawModule.references.map(
+      (ref: NapiModuleReference) => ({
+        module: queryModules[ref.index],
+        export: ref.export,
+        chunkingType: ref.chunkingType,
+      })
+    )
+    queryModule.incomingReferences = rawModule.incomingReferences.map(
+      (ref: NapiModuleReference) => ({
+        module: queryModules[ref.index],
+        export: ref.export,
+        chunkingType: ref.chunkingType,
+      })
+    )
+    modules.push(queryModule)
+  }
+  for (const entry of moduleGraph.entries) {
+    const queryModule = queryModules[entry]
+    entries.push(queryModule)
+  }
+}
+
+function runQuery(
+  query: string,
+  modules: Module[],
+  entries: Module[]
+): CallToolResult['content'] {
+  const response: CallToolResult['content'] = []
+  const proto = {
+    modules,
+    entries,
+    console: {
+      log: (...data: any[]) => {
+        response.push({
+          type: 'text',
+          text: data
+            .map((item) =>
+              typeof item === 'string' ? item : inspect(item, false, 2, false)
+            )
+            .join(' '),
+        })
+      },
+    },
+    global: undefined,
+    self: undefined,
+    globalThis: undefined,
+  }
+  const contextObject = Object.create(proto)
+  proto.global = contextObject
+  proto.self = contextObject
+  proto.globalThis = contextObject
```

Comment:
```suggestion
  contextObject.global = contextObject
  contextObject.self = contextObject
  contextObject.globalThis = contextObject
```
Should this be set on the object itself instead of modifying the prototype, or am I missing something?

Reply:
We could, but that doesn't make a big difference.